test/e2e/certificates_test.go

```diff
@@ -0,0 +1,366 @@
+
+			By("copying AWS secret access key to test namespace")
+			secretName := "aws-secret"
+			secretKey := "aws_secret_access_key"
+			awsSecret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: ns.Name,
+				},
+				Data: map[string][]byte{
+					secretKey: awsSecretAccessKey,
+				},
+			}
+			_, err = loader.KubeClient.CoreV1().Secrets(ns.Name).Create(ctx, awsSecret, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("getting AWS zone from Infrastructure object")
+			infra, err := infraClient.Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			awsRegion := infra.Status.PlatformStatus.AWS.Region
+			Expect(awsRegion).NotTo(Equal(""))
+
+			By("creating new certificate Issuer")
+			issuerName := "letsencrypt-dns01"
+			issuer := &certmanagerv1.Issuer{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      issuerName,
+					Namespace: ns.Name,
+				},
+				Spec: certmanagerv1.IssuerSpec{
+					IssuerConfig: certmanagerv1.IssuerConfig{
+						ACME: &v1.ACMEIssuer{
+							Server: "https://acme-staging-v02.api.letsencrypt.org/directory",
+							PrivateKey: certmanagermetav1.SecretKeySelector{
+								LocalObjectReference: certmanagermetav1.LocalObjectReference{
+									Name: "letsencrypt-dns01-issuer",
+								},
+							},
+							Solvers: []v1.ACMEChallengeSolver{
+								{
+									DNS01: &v1.ACMEChallengeSolverDNS01{
+										Route53: &v1.ACMEIssuerDNS01ProviderRoute53{
+											AccessKeyID: string(awsAccessKeyID),
+											SecretAccessKey: certmanagermetav1.SecretKeySelector{
+												LocalObjectReference: certmanagermetav1.LocalObjectReference{
+													Name: secretName,
+												},
+												Key: secretKey,
+											},
+											Region: awsRegion,
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			}
+			_, err = certmanagerClient.CertmanagerV1().Issuers(ns.Name).Create(ctx, issuer, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("creating new certificate")
+			certDomain := "adre." + appsDomain // acronym for "ACME dns-01 Route53 Explicit", short naming to pass dns name validation
+			certName := "letsencrypt-cert"
+			cert := &certmanagerv1.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      certName,
+					Namespace: ns.Name,
+				},
+				Spec: certmanagerv1.CertificateSpec{
+					IsCA:       false,
+					CommonName: certDomain,
+					SecretName: certName,
+					DNSNames:   []string{certDomain},
+					IssuerRef: certmanagermetav1.ObjectReference{
+						Name: issuerName,
+						Kind: "Issuer",
+					},
+				},
+			}
+
+			_, err = certmanagerClient.CertmanagerV1().Certificates(ns.Name).Create(ctx, cert, metav1.CreateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			By("checking for certificate secret contents")
+			err = wait.PollImmediate(PollInterval, TestTimeout, func() (bool, error) {
```

Comment:
This is intentional, checking there is no error returned by this function as PollImmediate gets pre-empted at the return of any error. In this case, the secrets get is the only call which is error-prone, however we expect for the first few minutes for this secret to not be present thus keep retyring with PollImmediate, but in the event of the returning the "secret not found" error from the Get call or the alike, the poll retries would stop which isn't desired.

Instead to avoid confusion, I can use `_ = wait.PollImmediate(...)` WDYT?

Reply:
`wait.PollImmediate` will return an error when the timeout occurs or the function of type `wait.ConditionFunc`returns false or an error. As you are not returning the error of "secret not found" from the function, it will return true or false based on the validity of the secret tls. Thus, the expected output is that the secret should be available within the timeout and it should be valid. Checking the error will validate this, or else we wouldn't know whether finally the correct secret was created or not.

Reply:
Thanks, I get your point (I think so), nevertheless realizing that PollImmediate would return an error even when timeout is exceeded, so always good to check the error out from it.
I will ensure to update all instances of err's returned from wait.PollImmediate to error check, after which you could take a re-look!

Reply:
All updated now :)